### ISSUE TITLE
Адмін модуль (UBOS)

### ISSUE THREAD

**Issue by pupenasan** [pupenasan/NodeREDGuidUKR#9]
### Група Authorization - призначена для 
- /login
- /signout
- /registration
- /repeatRegisterEmail
- /check_token

### Група Roles  - призначена для 
- /getRoles
- /getAllRoles
- /createRole
- /addModuleToRole
- /updateRole
- /deleteRole
- /deleteModuleFromRole

### Група Modules  - призначена для 
- /getModules
- /createParentModule
- /updateParentModule
- /deleteModule
- /createChildModule
- /updateChildModule
- /deleteChildModule

### Група User Management - призначена для 
- /getUsers
- /editUserRole
- /deleteUser
- /createUser

### Група Confirm Email Registration - призначена для підтвердження пошти користувача
- /verifyToken GET - підтвердження пошти користувача, передається token у вигляді параметра

### Група Reset Paaword - призначена для 
- /sendEmailForResetPassword
- /queryToResetPassword
- /resetPassword

**Comment by pupenasan** [comment]
У групі `Authorization` signout і registration мають однакову кінцеву точку.

**Comment by pupenasan** [comment]
@MashtalerOlha я планую розписувати даний модуль по деталям. У цій темі буду задавати питання і робити правку по загальній інформації.

**Comment by pupenasan** [comment]
Є /registration і /createUser. Чим відрізняється їх призначення?

**Comment by pupenasan** [comment]
Що таке Модулі?

**Comment by pupenasan** [comment]
> Є /registration і /createUser. Чим відрізняється їх призначення?

з UI все стає зрозуміліше. Це по ходу створення юзера адміном.

**Comment by MashtalerOlha** [comment]
> Є /registration і /createUser. Чим відрізняється їх призначення?

registration - створення користувача з верифікацією. Користувач реєструється самостійно
createUser -  без верифікації. Ми реєструємо користувача самостійно через адмін панель

**Comment by MashtalerOlha** [comment]
> орення юзер

так

**Comment by pupenasan** [comment]
Ось [тут](https://community.ubos.tech/olha/nalashtuvannia-avtorizatsiinogho-tiemplieitu-27la) написано
> І вставляємо перед /check_token посилання на свій Node Red.

Але там вже все вставлено автоматично.

**Comment by pupenasan** [comment]
Помилки в UI ігнорувати, так і має бути?

**Comment by pupenasan** [comment]
> ОБОВ'ЯЗКОВО id ролі user копіюємо та вставляємо у середовищі Node Red в function node - registration user

Мається на увазі `name` ролі?

**Comment by MashtalerOlha** [comment]
> Дії в методичці треба робити на опублікованій сторінці чи в середовищі розроблення?

які саме дії?

**Comment by MashtalerOlha** [comment]
> Помилки в UI ігнорувати, так і має бути?

які саме помилки?

**Comment by MashtalerOlha** [comment]
> 

id

**Comment by pupenasan** [comment]
> > 
> 
> id

у нас ніде не зустрічається id

**Comment by pupenasan** [comment]
> > Дії в методичці треба робити на опублікованій сторінці чи в середовищі розроблення?
> 
> які саме дії?

добавлення ролей а потім адміна

**Comment by pupenasan** [comment]
> > Помилки в UI ігнорувати, так і має бути?
> 
> які саме помилки?

та їх багато в мене, якщо у Вас немає, тоді заскріню кілька

**Comment by pupenasan** [comment]
> > > 
> > 
> > 
> > id
> 
> у нас ніде не зустрічається id

так, знайшов id, але це не очевидно, бо в таблиці ця колонка прихована, можливо принаймні написати де його подивитися

**Comment by pupenasan** [comment]
У потоці обробки `/login` у вузлі `check auth` помилка "Email not activated" ніколи не вийде, бо ми шукаємо тільки активних користувачів у `mongo Find`. Чи я чогось недобачив?

**Comment by pupenasan** [comment]
Це не appsmith налаштування, а Ubos UI так? Якщо так, то можна кілька слів про логіку роботи check auth, щоб розуміти 
коли він робить запит. 
![зображення](https://github.com/pupenasan/NodeREDGuidUKR/assets/25325920/8ce5a4d5-1774-4250-afa6-75bfaa2457b5)

**Comment by pupenasan** [comment]
Потік `POST /createChildModule`
```JS
sg.collection = "admin_modules";

//якщо вказаний батьківський модуль
if (msg.payload.parentModule) {
    msg.query={
        "_id": objectid(msg.payload.parentModule)
    };

    msg.payload = {
        $addToSet: { //Додає елементи до масиву, лише якщо вони ще не існують у наборі.
            // цей елемент буде додано до масиву
            childModules: {
                name: msg.payload.childModule,
                page_id: msg.payload.childModule,
                position: msg.payload.position
            }
        },
        $set: {
            isParent: true
        }
    }

    return [msg,];
} else {
    
    
    msg.payload={
        ...msg.payload,
        deleted: false,
        isParent: false
    }
    return [,msg];
}
```
Тобто дочірні модулі можна добавляти без бітьківського?

**Comment by pupenasan** [comment]
Судячи також з логіки, модуль стає батьківським, тільки коли в нього з'являються дочірні?

**Comment by pupenasan** [comment]
Якщо пробіли в назві сторінок писати і вказувати як модулі то він уходить в нескінченне намагання відкриватися

**Comment by pupenasan** [comment]
> Якщо пробіли в назві сторінок писати і вказувати як модулі то він уходить в нескінченне намагання відкриватися

З кирилицею та ж фігня. Треба щось буде з цим робити, бо меню мало б бути і українською доступне.

**Comment by pupenasan** [comment]
Також мабуть варто зробити вирівнювання в меню по правому краю бо ієрархічність погано відображається.

**Comment by pupenasan** [comment]
Не знайшов використання кінцевої точки `/updateRole`

**Comment by pupenasan** [comment]
Не знайшов де викликається сторінка `Info`
